File: SSUMobile/Modules/Directory/SSUDirectoryModule.m
//
//  SSUDirectoryModule.m
//  SSUMobile
//
//  Created by Eric Amorde on 9/8/15.
//  Copyright (c) 2015 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUDirectoryModule.h"
#import "SSUDirectoryBuilder.h"
#import "SSUMoonlightCommunicator.h"
#import "SSULogging.h"
#import "SSUConfiguration.h"
#import "SSUDirectorySpotlightUtilities.h"
#import "SSUDirectoryViewController.h"

@import CoreSpotlight;
@import MobileCoreServices;

@interface SSUDirectoryModule()

@end

@implementation SSUDirectoryModule

+ (instancetype) sharedInstance {
    static SSUDirectoryModule * instance = nil;
    static dispatch_once_t token;
    dispatch_once(&token, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

#pragma mark - SSUModule

- (nonnull NSString *) title {
    return NSLocalizedString(@"Directory",
                             @"The campus directory containing the contact information of faculty and staff");
}

- (nonnull NSString *) identifier {
    return @"directory";
}

- (UIView *) viewForHomeScreen {
    //return [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"directory_icon"]];
    UIImage * image = [UIImage imageNamed:@"directory_icon"];
    UIButton * button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button setImage:image forState:UIControlStateNormal];
    button.contentMode = UIViewContentModeScaleAspectFit;
    return button;
}

- (UIImage *) imageForHomeScreen {
    return [UIImage imageNamed:@"directory_icon"];
}

- (BOOL) showModuleInNavigationBar {
    return NO;
}

- (BOOL) shouldNavigateToModule {
    return YES;
}

- (void) selectHomeScreenView:(UIView *)view {
    UIButton * button = (UIButton *)view;
    [button setHighlighted:YES];
}

- (void) deselectHomeScreenView:(UIView *)view {
    UIButton * button = (UIButton *)view;
    [button setHighlighted:NO];
}

- (UIViewController *) initialViewController {
    UIStoryboard * storyboard = [UIStoryboard storyboardWithName:@"Directory_iPhone"
                                                          bundle:[NSBundle bundleForClass:[self class]]];
    return [storyboard instantiateInitialViewController];
}

- (void) setup {
    NSManagedObjectModel * model = [self modelWithName:@"Directory"];
    NSPersistentStoreCoordinator * coordinator = [self persistentStoreCoordinatorWithName:@"Directory" model:model];
    self.context = [self contextWithPersistentStoreCoordinator:coordinator];
    self.backgroundContext = [self backgroundContextFromContext:self.context];
}

- (void) clearCachedData {
    //TODO: implement clearCachedData
}

- (void) updateData:(void (^)())completion {
    SSULogDebug(@"Update Directory");
    
    [self updateBuildings:^{
        [self updateSchools:^{
            [self updateDepartments:^{
                [self updatePeople:^{
                    if (completion != NULL) {
                        completion();
                    }
                    [SSUDirectorySpotlightUtilities populateIndex:[CSSearchableIndex defaultSearchableIndex] context:self.backgroundContext domain:nil];
                }];
            }];
        }];
    }];
}

- (void) updatePeople:(void(^)())completion {
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUDirectoryPersonUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"directory/person/" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update directory person: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self buildPerson:json completion:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUDirectoryPersonUpdatedDateKey];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

- (void) updateDepartments:(void(^)())completion {
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUDirectoryDepartmentUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"directory/department/" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update directory departments: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self buildDepartment:json completion:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUDirectoryDepartmentUpdatedDateKey];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

- (void) updateSchools:(void(^)())completion {
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUDirectorySchoolUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"directory/school/" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update directory schools: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self buildSchool:json completion:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUDirectorySchoolUpdatedDateKey];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

- (void) updateBuildings:(void(^)())completion {
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUDirectoryBuildingUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"directory/building/" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update directory buildings: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self buildBuilding:json completion:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUDirectoryBuildingUpdatedDateKey];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

#pragma mark - Private

- (void) buildPerson:(NSArray *)personData completion:(void(^)(void))completion {
    SSUDirectoryBuilder * builder = [[SSUDirectoryBuilder alloc] init];
    builder.context = [self newBackgroundContext];
    [builder.context performBlock:^{
        [builder buildPeople:personData];
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion();
            });
        }
    }];
}

- (void) buildDepartment:(NSArray *)departmentData completion:(void(^)(void))completion {
    SSUDirectoryBuilder * builder = [[SSUDirectoryBuilder alloc] init];
    builder.context = [self newBackgroundContext];
    [builder.context performBlock:^{
        [builder buildDepartments:departmentData];
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion();
            });
        }
    }];
}

- (void) buildSchool:(NSArray *)schoolData completion:(void(^)(void))completion {
    SSUDirectoryBuilder * builder = [[SSUDirectoryBuilder alloc] init];
    builder.context = [self newBackgroundContext];
    [builder.context performBlock:^{
        [builder buildSchools:schoolData];
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion();
            });
        }
    }];
}

- (void) buildBuilding:(NSArray *)buildingData completion:(void(^)(void))completion {
    SSUDirectoryBuilder * builder = [[SSUDirectoryBuilder alloc] init];
    builder.context = [self newBackgroundContext];
    [builder.context performBlock:^{
        [builder buildBuildings:buildingData];
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion();
            });
        }
    }];
}

#pragma mark - Spotlight

- (void) searchableIndex:(CSSearchableIndex *)index reindexItemWithIdentifier:(NSString *)identifier {
    [SSUDirectorySpotlightUtilities searchableIndex:index reindexItem:identifier inContext:self.backgroundContext domain:self.identifier];
}

- (void) searchAbleIndexRequestingUpdate:(CSSearchableIndex *)index {
    [SSUDirectorySpotlightUtilities populateIndex:index context:self.backgroundContext domain:self.identifier];
}

- (BOOL) recognizesIdentifier:(NSString *)identifier {
    return YES;
}

- (UIViewController *) viewControllerForSearchableItemWithIdentifier:(NSString *)identfier {
    SSUDirectoryViewController * vc = [SSUDirectoryViewController instantiateFromStoryboard];
    vc.objectToDisplay = [SSUDirectorySpotlightUtilities objectForIdentifier:identfier];
    return vc;
}

@end

File: SSUMobile/Modules/Directory/SSUDirectoryModule.swift
//
//  SSUDirectoryModule.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/9/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation
import CoreSpotlight

class SSUDirectoryModule: SSUCoreDataModuleBase, SSUModuleUI, SSUSpotlightSupportedProtocol {
    
    @objc(sharedInstance)
    static let instance = SSUDirectoryModule()
    
    // MARK: SSUModule
    
    var title: String {
        return NSLocalizedString("Directory", comment: "The campus directory containing the contact information of faculty and staff")
    }
    
    var identifier: String {
        return "directory"
    }
    
    func setup() {
        setupCoreData(modelName: "Directory", storeName: "Directory")
    }
    
    func updateData(_ completion: (() -> Void)? = nil) {
        SSULogging.logDebug("Update Directory")
        updateBuildings {
            self.updateSchools {
                self.updateDepartments {
                    self.updateBuildings {
                        completion?()
                    }
                }
            }
        }
    }
    
    
    func updatePeople(completion: (() -> Void)? = nil) {
        let date = SSUConfiguration.sharedInstance().date(forKey:SSUDirectoryPersonUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("directory/person", since:date) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Directory People: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUDirectoryPersonUpdatedDateKey)
                self.buildPeople(json) {
                    completion?()
                }
            }
        }
    }
    
    func updateDepartments(completion: (() -> Void)? = nil) {
        let date = SSUConfiguration.sharedInstance().date(forKey:SSUDirectoryDepartmentUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("directory/deparment", since:date) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Directory departments: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUDirectoryDepartmentUpdatedDateKey)
                self.buildDepartment(json) {
                    completion?()
                }
            }
        }
    }
    
    func updateBuildings(completion: (() -> Void)? = nil) {
        let date = SSUConfiguration.sharedInstance().date(forKey:SSUDirectoryBuildingUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("directory/building", since:date) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Directory buildings: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUDirectoryBuildingUpdatedDateKey)
                self.buildBuildings(json) {
                    completion?()
                }
            }
        }
    }

    func updateSchools(completion: (() -> Void)? = nil) {
        let date = SSUConfiguration.sharedInstance().date(forKey:SSUDirectorySchoolUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("directory/school", since:date) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Directory schools: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUDirectorySchoolUpdatedDateKey)
                self.buildSchools(json) {
                    completion?()
                }
            }
        }
    }
    
    
    func buildPeople(_ json: Any, completion: (() -> Void)? = nil) {
        guard let data = json as? [Any] else {
            completion?()
            return
        }
        let builder = SSUDirectoryBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.buildPeople(data)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    func buildDepartment(_ json: Any, completion: (() -> Void)? = nil) {
        guard let data = json as? [Any] else {
            completion?()
            return
        }
        let builder = SSUDirectoryBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.buildDepartments(data)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    func buildSchools(_ json: Any, completion: (() -> Void)? = nil) {
        guard let data = json as? [Any] else {
            completion?()
            return
        }
        let builder = SSUDirectoryBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.buildSchools(data)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    func buildBuildings(_ json: Any, completion: (() -> Void)? = nil) {
        guard let data = json as? [Any] else {
            completion?()
            return
        }
        let builder = SSUDirectoryBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.buildBuildings(data)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    // MARK: SSUModuleUI
    
    func imageForHomeScreen() -> UIImage? {
        return UIImage(named: "directory_icon")
    }
    
    func viewForHomeScreen() -> UIView? {
        return nil
    }
    
    func initialViewController() -> UIViewController {
        let storyboard = UIStoryboard(name: "Directory_iPhone", bundle: Bundle(for: type(of: self)))
        return storyboard.instantiateInitialViewController()!
    }
    
    func shouldNavigateToModule() -> Bool {
        return true
    }
    
    func showModuleInNavigationBar() -> Bool {
        return false
    }
    
    // MARK: Spotlight
    
    @available(iOS 9.0, *)
    func searchableIndex(_ index: CSSearchableIndex!, reindexItemWithIdentifier identifier: String!) {
        SSUDirectorySpotlightUtilities.searchableIndex(index, reindexItem: identifier, in: backgroundContext, domain: self.identifier)
    }
    
    @available(iOS 9.0, *)
    func searchAbleIndexRequestingUpdate(_ index: CSSearchableIndex!) {
        SSUDirectorySpotlightUtilities.populateIndex(index, context: backgroundContext, domain: self.identifier)
    }
    
    func recognizesIdentifier(_ identifier: String!) -> Bool {
        return identifier.hasPrefix(self.identifier)
    }
    
    func viewControllerForSearchableItem(withIdentifier identfier: String!) -> UIViewController! {
        let vc = SSUDirectoryViewController.instantiateFromStoryboard()
        vc?.objectToDisplay = SSUDirectorySpotlightUtilities.object(forIdentifier: identifier)
        return vc
    }
}
